### ISSUE TITLE
Question about  EMA

### ISSUE THREAD

**Issue by BaohaoLiao** [Yuanhy1997/SeqDiffuSeq#8]
Hi,

I'm confused about the EMA implementation. Could you tell me whether my understanding is correct? 

There are two ways to do EMA. The first one is
```
1. after optimizer.step() you obtain model_new
2. ema_model_new = a * ema_model_old + (1-a) * model_new   # a is ema_decay
3. model_new = copy(ema_model_new)
```
The second one is:
```
1. after optimizer.step() you obtain model_new
2. ema_model_new = a * ema_model_old + (1-a) * model_new   # a is ema_decay
```
The difference between these two ways are:

- the first one needs to use the new ema_model parameters to initialize the model for the next step. Till convergence, ema_model and model should be very similar. We can use either to do sampling.
- The second one just wants to store the ema_model and uses it for sampling. The update of ema_model will not affect the model.

May I ask which one you implement? I assume it is the second one, since the training outputs don't change when I change ema_decay.

**Comment by Yuanhy1997** [comment]
I think it the second one. It is pretty clear if you read the code of trainer class (the optimizer in the init function and the normal optimize function).

**Comment by BaohaoLiao** [comment]
Thank you, just want to confirm it. 

In addition, may I ask why you use alpha_cumprod_step_260000.npy for sampling? I understand you save the loss history in it for resuming training if one uses a  loss-aware scheduler. However, alpha_cumprod is fixed and you can recompute it for sampling instead of loading it from an external file. May I know your motivation for saving it?

**Comment by Yuanhy1997** [comment]
I don't really get your question. We save the loss history file for some analysis use and not intended for resuming training. Alpha cumprod is not fixed during training and is our access to the noise schedule. It directly controls the level of noises and from alpha cumprod we can get the alpha and beta by some simple calculation. We modify the alpha cumprod every 20000 steps out of the consideration of stable updating. You may change the hyper param to other values.

**Comment by BaohaoLiao** [comment]
sorry, I forgot your adaptive noising part and considered it as statically noising. Thank you for your help.
